### ISSUE TITLE
Use a single project lock for dependency management and task execution

### ISSUE THREAD

**Issue by ghale** [gradle/gradle-native#830]
The IDE performance tests have been experiencing intermittent build failures since we introduced the checks for when a domain object has been removed from its container:

https://builds.gradle.org/repository/download/Gradle_Check_IndividualPerformanceScenarioWorkersLinux/15099801:id/failure-logs/performance/build/tmp/performance-test-files/IdeIntegrationPerformanceTest/idea_on_largeJavaMultiProject/76s3k/current________________/log.txt

It's unlikely that these tasks have actually been removed and its more likely that this is related to a concurrency issue where the object appears as if it has been removed for a small window during the process of creation and another thread is accessing the object during this window.

We need to investigate this and address the problem.

**Comment by lacasseio** [comment]
This error doesn't make sense so far as the error happens at execution time. Unless I'm miss understanding how the task graph is constructed, by the time Gradle reach execution time, it should have already realized all the task required for execution. In this case, it seems that parallel execution is creating all the condition for 2 thread to be realizing the same `TaskProvider`. In this case, the object is constructed but not yet added to the container which triggers the condition where the provider is removed from the container. We could change this condition but we would simply hide the error by allowing a possibility of 2 tasks to be created which would cause another set of issues. The resolution of the task can happen in only 3 places:
- Via a convention mapping: https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java#L320
- Via the dependency manager: https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/internal/IdeaModuleMetadata.java#L48
- Via task graph generation

Let's validate the following assumption:

- All tasks in the execution graph are already realized.
- The task graph is built before the execution phase starts.
- The incubating parallel feature executes projects in parallel but not the tasks.
- Only a single creating provider should be created when registering a task.
- The dependency management locks appropriately when reaching between projects.
- Task realization happens only through the creating provider.

If any of the previous assumptions are wrong, then we could probably explain the issue. I welcome any additional solution path to explore.

**Comment by lacasseio** [comment]
A workaround fix would be: https://github.com/gradle/gradle/compare/gh/lazy-tasks/idea-removal-error

I believe this would just hide the issue but would avoid CI to fail.

**Comment by lacasseio** [comment]
PR https://github.com/gradle/gradle/pull/6443 seems to have also triggered this failure, let's use it as another data point.

**Comment by ghale** [comment]
Actually, I think we potentially have two issues here.  First, we know we have a convention mapping in this plugin that could realize a task in another project at execution time.  For instance, you could have project A and project B that both have an IDE artifact dependency on project C.  If the `ideaModule` tasks for A and B run concurrently (and C is not in the task graph) then they could attempt to realize C simultaneously, so we need to use the project mutation lock to prevent that from happening.  Second, in this particular build, we are running `ideaModule` for the whole multi-project build, therefore we expect the `ideaModule` task from every project to be realized when the execution graph is built.  It doesn't make sense for it not to be realized at execution time.  The fact that we see it being realized during execution implies that there is something else weird going on - maybe multiple tasks in the container?  The fix for the first would prevent an error in this case, but there could still be an underlying issue here.  To your point, @lacasseio, we should get to the bottom of this second issue before we fix the first.

**Comment by ghale** [comment]
Ok, I've done a little more research into this and I think I have a theory for the behavior we are seeing.  First of all, the convention mappings never get called from other projects - e.g. the convention mapping for the project183 ideaModule will only get called from the `:project183:ideaModule` task and not from another project.  Second, when we resolve the IDE artifact dependencies, we typically do create some tasks - for instance, jar, processResources, and/or compileJava tasks in the dependency project.  This creation is done while holding the project mutation lock, so it is safe from itself (for instance project A and B both resolving dependencies in the same dependency project at the same time).  However, the convention mapping `get()` does not use the mutation lock, so it can jump in at any time (potentially when another project is resolving dependencies that include the project where the convention mapping is being called).  My suspicion is that the `ideaModule` task is not being created multiple times, but that we are checking the removed status on the task while another task in that project is in the process of being created (for instance, processResources as a result of IDE artifact resolution).  The `Map` that is used in the index in `NamedDomainObjectCollection` is not thread-safe and can do rebalancing after insertion, which could affect a simultaneous `get()` on the map.  I suspect we are checking for the existence of the task in the `Map` at the same time that `put()` on another task is happening, and in that instant the `ideaModule` task appears like it has been removed (i.e. `object != null` but the task name does not appear in the index).  Given that, I think simply fixing the convention mapping so that it uses the project mutation lock will clean up the issue.

**Comment by adammurdoch** [comment]
If a convention mapping needs to deal with concurrency, then we're doing it wrong. Convention mappings are just another part of the project state, so should always be accessed from a single thread at a time. If there is something calling a convention mapping on a project while another thread is doing something else with that project, then we're missing some synchronisation, but at a much higher level than the convention mappings. We should instead figure out why these 2 threads are touching the project at the same time and fix that.

**Comment by ghale** [comment]
Here's a WIP spike of using a single lock for mutating project state: https://github.com/gradle/gradle/pull/6475

Comments/feedback welcome.

**Comment by ghale** [comment]
So, there are some issues with joining the project locks that I'm trying to navigate.  When a task is executing, we hold the project lock for the project that task is in (let's call that project A).  If we have to resolve an artifact in another project, we grab the project lock for that project (let's call that project B).  If we continue to hold the lock on project A, we could create a deadlock condition if a task in project B now decides to resolve an artifact in project A.  So in my spike, we release the project A lock, grab the project B lock, resolve it's artifact, and then reacquire the project A lock before proceeding.  This prevents a deadlock of the aforementioned type.  

However, there is a synchronization lock around `DefaultResolvedArtifact.getFile()` (https://github.com/gradle/gradle/blob/1993fd0e3fa9e8c9d0f1b020bab9fb4d5bd4512a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultResolvedArtifact.java#L134) which creates another deadlock situation.  What happens is that when we release the project lock for the project this task is executing in, another task in the same project can start running.  If that task (now holding the project lock) also tries to resolve the same artifact, he can block acquiring the synchronization lock (meanwhile the other thread is blocked inside the synchronized code trying to reacquire the project lock).

A couple of options:

- Move the synchronization for the artifact resolution down into the resolvers.  So, `ProjectDependencyResolver` currently protects against simultaneous resolution of the same artifact by locking the project, but other resolvers would need to implement some sort of synchronization to ensure that, for instance, we don't download the same artifact from a repository multiple times.
- Make the synchronization around the artifact some sort of `ResourceLock` that can be released and reacquired atomically with the project lock.
- Rather than treat the two locks equivalently, create a hierarchical relationship where a task can release mutability access to a project without releasing execution access.  In other words, a task must acquire both execution and mutability access to its project before it can run.  It can then temporarily release mutability access when accessing other projects, which would allow tasks in other projects to access this project, but another task in the same project cannot run until the first task releases both execution and mutability access.

I think the last option is what I'm leaning towards.  There might be other approaches I'm not thinking about, though.  Thoughts?

**Comment by adammurdoch** [comment]
The lock in `DefaultResolvedArtifact` is only protecting the state of the instance it's attached to. It's not there to coordinate access to the resources (that's already in place further down).

Give that `DefaultResolvedArtifact` is a model element, we shouldn't need any synchronisation in that type at all. I'd remove the lock.

**Comment by ghale** [comment]
Another issue we're going to have to deal with here is that this change breaks the gretty plugin.  This plugin launches an unsafe thread that then resolves a configuration.  Meanwhile, the parent task waits on this thread (holding the project lock and creating a deadlock).  This should be manageable by turning this into a worker api call or something, but this is a relatively popular plugin so we'll need to engage the owner and potentially contribute a PR.

**Comment by ghale** [comment]
Per our discussions, I've removed the places where I had added implicitly locking in the configuration object, so that now, during configuration resolution, we only lock where we were doing so before (when we step over to another project) and we throw a deprecation error if something tries to resolve a configuration without have the lock.  However, the problem with the "resolves dependencies from an external repo" did not go away.  I dug into this build and I now have a pretty clear idea of what's going on and I'm even more convinced that it has to do with the coarser locking.

In [this example](https://github.com/gradle/performance-comparisons/tree/master/parallel-downloads), we have 2 projects: application and library.  Each has dependencies on some external stuff, application has a dependency on library, and both apply the "io.spring.dependency-management" plugin.  At execution time, they each have a task that sequentially resolves their compile, testCompile, runtime and testRuntime configurations in the task action.  

At execution time, the application task fires up and starts resolving his configurations, meanwhile the library task does the same.  When the application task gets to the library dependency, he attempts to lock the library project before resolving the configuration in that project.  The library task, however, is holding that lock, so the application task has to wait until the library task is finished.  Additionally, the spring dependency management plugin registers a "beforeResolve" hook on the configurations that in turn creates and resolves additional detached configurations (see [here](https://github.com/spring-gradle-plugins/dependency-management-plugin/blob/master/src/main/java/io/spring/gradle/dependencymanagement/internal/ExclusionConfiguringAction.java#L110-L140)).  Prior to the changes to use a single project lock, the application task could slip in anywhere in between those configurations and resolve the library configuration before continuing, but now it has to wait for all of the configurations that the task resolves (testCompile, runtime and testRuntime and all of the extra detached configurations) to all resolve before it can continue.  In fact, the difference in task execution times between the two is almost exactly the time spent resolving the additional configurations.  Much of the resolution in the consuming project could run in parallel with the blocking work in the other project, but it doesn't because it gets blocked waiting for access to the other project.

I think this is expected behavior, but it's aggravated in this case because of the additional work being done in the other project while holding the project lock.

**Comment by adammurdoch** [comment]
The problem sounds like we're holding a lock around looking up the in-memory project metadata caches. Here's how it should work:

To run the 'library' task:

1. grab the lock on 'library' project
2. start running the 'library' task
3. start to resolve the first configuration
4. resolution calculates the metadata for 'library' and shoves it in the in-memory cache.
5. the task continues ...

Meanwhile, to run the 'application' task:

1. grab the lock on 'application' project
2. start running the 'application' task.
3. start to resolve the first configuration
4. to resolve the dependency on the 'library' project, the thread looks in the in-memory cache. It _does not_ need to acquire the lock on 'library' to do this. Only if the meta-data is missing does it need to lock the target project in order to create the metadata. Only one thread needs to do this
5. the task continues ...

The same thing should happen with resolving the artefacts. Only if the file location has not been calculated from the project state does the lock on the project need to be acquired. Otherwise, the result can be used directly out of the cached metadata.

**Comment by ghale** [comment]
Here's a PR for some preliminary work here: https://github.com/gradle/gradle/pull/6861

@big-guy can you take a look?

**Comment by ghale** [comment]
This PR has the meat of the changes: https://github.com/gradle/gradle/pull/6886

@big-guy Can you take a look?

**Comment by big-guy** [comment]
@ghale Thinking a bit more, it might be easier to retarget this to release and just wait for the milestone to be cut tomorrow. I think we'll need a new nightly to rebaseline the performance regressions.  

Can you open a new issue for `markThisObserved`? We can close this issue once the PR is on release and the perf tests are rebaselined.
